# HÆ°á»›ng Dáº«n Tá»• Chá»©c File & Folder

> **Má»¥c Ä‘Ã­ch**: Quy Ä‘á»‹nh rÃµ rÃ ng file nÃ o náº±m folder nÃ o, trÃ¡nh lá»™n xá»™n

---

## ğŸ“ Cáº¥u TrÃºc ThÆ° Má»¥c

```
GameBuild/
â”œâ”€â”€ ğŸ“„ index.html                 # Entry point cá»§a game
â”œâ”€â”€ ğŸ“„ PROJECT_STATUS.md          # Tráº¡ng thÃ¡i dá»± Ã¡n (ROOT level)
â”œâ”€â”€ ğŸ“„ README.md                  # Tá»•ng quan dá»± Ã¡n
â”œâ”€â”€ ğŸ“„ .gitignore                 # Git ignore rules
â”‚
â”œâ”€â”€ ğŸ“ src/                       # MÃƒ NGUá»’N PRODUCTION
â”‚   â”œâ”€â”€ ğŸ“ core/                  # Game engine core
â”‚   â”‚   â”œâ”€â”€ game.js              # Game loop, orchestration
â”‚   â”‚   â”œâ”€â”€ state.js             # State management
â”‚   â”‚   â””â”€â”€ events.js            # Event system
â”‚   â”‚
â”‚   â”œâ”€â”€ ğŸ“ ai/                    # TÃCH Há»¢P AI
â”‚   â”‚   â”œâ”€â”€ gemini.js            # Gemini API integration
â”‚   â”‚   â”œâ”€â”€ context.js           # Context builder
â”‚   â”‚   â”œâ”€â”€ parser.js            # Response parser
â”‚   â”‚   â””â”€â”€ local-ai.js          # Local AI (Phase 3+)
â”‚   â”‚
â”‚   â””â”€â”€ ğŸ“ ui/                    # USER INTERFACE
â”‚       â”œâ”€â”€ chat.js              # Chat interface logic
â”‚       â”œâ”€â”€ stats.js             # Stats panel
â”‚       â””â”€â”€ renderer.js          # Display rendering
â”‚
â”œâ”€â”€ ğŸ“ data/                      # Dá»® LIá»†U GAME
â”‚   â”œâ”€â”€ ğŸ“ prompts/              # System prompts cho AI
â”‚   â”‚   â”œâ”€â”€ game-master.md       # Main GM prompt
â”‚   â”‚   â”œâ”€â”€ npc-personalities.md # NPCs prompts
â”‚   â”‚   â””â”€â”€ world-lore.md        # World context
â”‚   â”‚
â”‚   â””â”€â”€ ğŸ“ static/               # Static content (náº¿u cÃ³)
â”‚       â””â”€â”€ initial-scene.json   # Opening scene
â”‚
â”œâ”€â”€ ğŸ“ assets/                    # TÃ€I NGUYÃŠN TÄ¨NH
â”‚   â”œâ”€â”€ ğŸ“ css/                  # Styles
â”‚   â”‚   â”œâ”€â”€ main.css             # Main stylesheet
â”‚   â”‚   â”œâ”€â”€ chat.css             # Chat UI
â”‚   â”‚   â””â”€â”€ theme.css            # Color theme
â”‚   â”‚
â”‚   â”œâ”€â”€ ğŸ“ images/               # HÃ¬nh áº£nh (náº¿u cÃ³)
â”‚   â””â”€â”€ ğŸ“ sounds/               # Ã‚m thanh (Phase 2+)
â”‚
â”œâ”€â”€ ğŸ“ docs/                      # TÃ€I LIá»†U
â”‚   â”œâ”€â”€ ğŸ“ architecture/         # Kiáº¿n trÃºc ká»¹ thuáº­t
â”‚   â”‚   â”œâ”€â”€ ARCHITECTURE.md      # Full architecture (Phase 3+)
â”‚   â”‚   â”œâ”€â”€ MVP_ARCHITECTURE.md  # MVP approach â­
â”‚   â”‚   â”œâ”€â”€ CONTRACTS.md         # System contracts
â”‚   â”‚   â””â”€â”€ AI_INTEGRATION.md    # AI integration guide
â”‚   â”‚
â”‚   â”œâ”€â”€ ğŸ“ guides/               # HÆ°á»›ng dáº«n
â”‚   â”‚   â”œâ”€â”€ PROMPTING.md         # Prompt engineering
â”‚   â”‚   â””â”€â”€ DEPLOYMENT.md        # Deploy guide
â”‚   â”‚
â”‚   â”œâ”€â”€ ğŸ“ rules/                # Quy táº¯c dá»± Ã¡n
â”‚   â”‚   â”œâ”€â”€ update-status.md     # Update status rule
â”‚   â”‚   â””â”€â”€ file-organization.md # File nÃ y!
â”‚   â”‚
â”‚   â”œâ”€â”€ DEVELOPMENT_RULES.md     # Coding standards
â”‚   â””â”€â”€ CRITICAL_ISSUES.md       # Issues tracker
â”‚
â”œâ”€â”€ ğŸ“ ideas/                     # Ã TÆ¯á»NG & BRAINSTORM
â”‚   â””â”€â”€ game-concepts.md         # Game concepts
â”‚
â”œâ”€â”€ ğŸ“ test/                      # TEST CODE âš ï¸
â”‚   â”œâ”€â”€ ğŸ“ unit/                 # Unit tests
â”‚   â”œâ”€â”€ ğŸ“ integration/          # Integration tests
â”‚   â””â”€â”€ ğŸ“ playground/           # Thá»­ nghiá»‡m
â”‚   â””â”€â”€ âš ï¸ XÃ“A FOLDER NÃ€Y TRÆ¯á»šC KHI RELEASE!
â”‚
â””â”€â”€ ğŸ“ .agent/                    # Agent workflows (gitignored)
    â””â”€â”€ workflows/
        â””â”€â”€ cleanup-test-files.md
```

---

## ğŸ—‚ï¸ Quy Táº¯c Äáº·t File

### 1. **Code Production** â†’ `src/`

**NguyÃªn táº¯c**:
- âœ… Code cháº¡y thá»±c táº¿
- âœ… ÄÆ°á»£c import trong game
- âŒ KHÃ”NG cÃ³ test code
- âŒ KHÃ”NG cÃ³ experimental code

**VÃ­ dá»¥**:
```javascript
// âœ… ÄÃšNG: src/ai/gemini.js
export async function callGemini(prompt) { ... }

// âŒ SAI: src/ai/test-gemini.js
function testGeminiAPI() { ... }
```

---

### 2. **AI-Related Code** â†’ `src/ai/`

**Khi nÃ o dÃ¹ng**:
- âœ… API integration (Gemini, OpenAI)
- âœ… Context building
- âœ… Response parsing
- âœ… Local AI models

**Files**:
```
src/ai/
â”œâ”€â”€ gemini.js          # Gemini API wrapper
â”œâ”€â”€ context.js         # Build context for AI
â”œâ”€â”€ parser.js          # Parse AI responses
â””â”€â”€ local-ai.js        # Local models (Phase 3+)
```

---

### 3. **System Prompts** â†’ `data/prompts/`

**Äá»‹nh dáº¡ng**: Markdown (`.md`)

**Quy táº¯c**:
- âœ… Má»—i prompt 1 file riÃªng
- âœ… Dá»… edit, khÃ´ng cáº§n rebuild
- âœ… Version control friendly

**VÃ­ dá»¥**:
```markdown
<!-- data/prompts/game-master.md -->
You are the Game Master for "The Last Voyage"...

Rules:
- ...
```

---

### 4. **UI/Styling** â†’ `assets/css/`

**Quy táº¯c**:
- âœ… TÃ¡ch file theo component
- âœ… Main theme trong `theme.css`
- âŒ KhÃ´ng inline CSS trong HTML

**Files**:
```
assets/css/
â”œâ”€â”€ main.css           # Global styles, imports
â”œâ”€â”€ chat.css           # Chat interface
â”œâ”€â”€ stats.css          # Stats panel
â””â”€â”€ theme.css          # Colors, fonts, variables
```

---

### 5. **Documentation** â†’ `docs/`

**PhÃ¢n loáº¡i**:

| Loáº¡i | Folder | VÃ­ dá»¥ |
|------|--------|-------|
| **Architecture** | `docs/architecture/` | ARCHITECTURE.md |
| **Guides** | `docs/guides/` | PROMPTING.md |
| **Rules** | `docs/rules/` | update-status.md |
| **Root docs** | `docs/` | DEVELOPMENT_RULES.md |

**NguyÃªn táº¯c Ä‘áº·t tÃªn**:
- UPPER_CASE.md cho docs quan trá»ng
- lower-case.md cho guides

---

### 6. **Test Code** â†’ `test/` âš ï¸

**QUAN TRá»ŒNG**:
- âœ… Táº¤T Cáº¢ test code vÃ o Ä‘Ã¢y
- âœ… Experimental code vÃ o `test/playground/`
- âš ï¸ **XÃ“A toÃ n bá»™ folder nÃ y trÆ°á»›c release!**

**Workflow**:
```bash
# Khi cáº§n xÃ³a test code
/cleanup-test-files
```

---

### 7. **Ideas & Brainstorm** â†’ `ideas/`

**Má»¥c Ä‘Ã­ch**:
- ğŸ“ Brainstorming
- ğŸ’¡ Game concepts
- ğŸ¨ Design ideas
- âŒ KHÃ”NG pháº£i docs chÃ­nh thá»©c

---

## ğŸ“‹ Checklist TrÆ°á»›c Khi Táº¡o File Má»›i

### Tá»± há»i:

1. **File nÃ y lÃ  loáº¡i gÃ¬?**
   - Code production â†’ `src/`
   - AI integration â†’ `src/ai/`
   - System prompt â†’ `data/prompts/`
   - CSS â†’ `assets/css/`
   - Documentation â†’ `docs/`
   - Test â†’ `test/`

2. **CÃ³ cáº§n folder con má»›i khÃ´ng?**
   - Náº¿u cÃ³ > 5 files cÃ¹ng loáº¡i â†’ Táº¡o subfolder

3. **Naming convention Ä‘Ãºng chÆ°a?**
   - Code: `camelCase.js`
   - Docs: `UPPER_CASE.md` hoáº·c `lower-case.md`
   - Prompts: `descriptive-name.md`

---

## ğŸ” VÃ­ Dá»¥ Thá»±c Táº¿

### â“ "TÃ´i muá»‘n táº¡o file Ä‘á»ƒ lÆ°u prompt cho NPC Marcus"

**âœ… ÄÃºng**: `data/prompts/npc-marcus.md`

**âŒ Sai**:
- `src/prompts/marcus.md` - prompts khÃ´ng pháº£i code
- `marcus-prompt.txt` - khÃ´ng follow structure
- `data/marcus.md` - thiáº¿u subfolder

---

### â“ "TÃ´i muá»‘n táº¡o file test cho Gemini API"

**âœ… ÄÃºng**: `test/integration/test-gemini-api.js`

**âŒ Sai**:
- `src/ai/test-gemini.js` - test khÃ´ng vÃ o src/
- `gemini-test.js` - khÃ´ng cÃ³ folder structure

---

### â“ "TÃ´i muá»‘n lÆ°u CSS cho chat interface"

**âœ… ÄÃºng**: `assets/css/chat.css`

**âŒ Sai**:
- `src/ui/chat.css` - CSS khÃ´ng vÃ o src/
- `chat-styles.css` - thiáº¿u folder structure
- Inline trong HTML - vi pháº¡m separation

---

## ğŸš¨ Red Flags - Dáº¥u Hiá»‡u Sai

### âŒ Náº¿u tháº¥y nhá»¯ng Ä‘iá»u nÃ y â†’ SAI!

```
âŒ src/test-something.js          # Test code trong src/
âŒ data/game.js                    # Code trong data/
âŒ marcus.md (root level)          # File láº» á»Ÿ root
âŒ src/styles.css                  # CSS trong src/
âŒ ai-prompt.txt                   # Prompt khÃ´ng cÃ³ structure
```

---

## âœ… Best Practices

### 1. **TÃ¡ch Biá»‡t RÃµ RÃ ng**
```
Code (src/) â‰  Data (data/) â‰  Docs (docs/) â‰  Test (test/)
```

### 2. **Má»™t File Má»™t Má»¥c ÄÃ­ch**
```javascript
// âœ… GOOD: ai/gemini.js chá»‰ lÃ m Gemini API
// âŒ BAD: ai/everything.js lÃ m táº¥t cáº£
```

### 3. **Naming Consistency**
```
âœ… chat.js, chat.css, chat.md      # Consistent naming
âŒ chat.js, chatUI.css, Chat.md    # Inconsistent
```

### 4. **Cleanup Regularly**
```bash
# Má»—i tuáº§n
1. Review test/ folder
2. XÃ³a experimental code
3. Move stable code to src/
```

---

## ğŸ“Š Quick Reference

| TÃ´i cáº§n... | Táº¡o á»Ÿ Ä‘Ã¢y | TÃªn file |
|-----------|-----------|----------|
| **AI API integration** | `src/ai/` | `gemini.js` |
| **System prompt** | `data/prompts/` | `game-master.md` |
| **Game logic** | `src/core/` | `game.js` |
| **UI rendering** | `src/ui/` | `renderer.js` |
| **Styling** | `assets/css/` | `main.css` |
| **Architecture doc** | `docs/architecture/` | `DESIGN.md` |
| **Guide** | `docs/guides/` | `setup.md` |
| **Test code** | `test/` | `test-*.js` |
| **Brainstorm** | `ideas/` | `concept.md` |

---

## ğŸ”„ Khi Cáº§n Refactor

**Dáº¥u hiá»‡u cáº§n refactor structure**:
- âŒ CÃ³ > 10 files trong 1 folder (khÃ´ng cÃ³ subfolder)
- âŒ KhÃ´ng biáº¿t file nÃ o náº±m Ä‘Ã¢u
- âŒ Test code láº«n vá»›i production code
- âŒ CSS scattered everywhere

**Giáº£i phÃ¡p**:
1. Táº¡o subfolder theo tÃ­nh nÄƒng
2. Move files vá» Ä‘Ãºng chá»—
3. Update imports
4. Test láº¡i

---

**Nhá»›**: Structure tá»‘t = Dá»… maintain = Ãt bug = Happy coding! ğŸ‰
